packages/common-scripts/src/p2pkh.ts

```diff
@@ -0,0 +1,136 @@
+import { Cell, core, utils } from "@ckb-lumos/base";
+import {
+  TransactionSkeletonType,
+  createTransactionFromSkeleton,
+} from "@ckb-lumos/helpers";
+import { Reader, normalizers } from "@ckb-lumos/toolkit";
+import { Hash, Script } from "@ckb-lumos/base";
+import { BI } from "@ckb-lumos/bi";
+
+function groupInputs(inputs: Cell[], locks: Script[]): Map<string, number[]> {
+  const lockSet = new Set<string>();
+  for (const lock of locks) {
+    const scriptHash = utils
+      .ckbHash(core.SerializeScript(normalizers.NormalizeScript(lock)))
+      .serializeJson();
+    lockSet.add(scriptHash);
+  }
+
+  const groups = new Map<string, number[]>();
+  for (let i = 0; i < inputs.length; i++) {
+    const scriptHash = utils
+      .ckbHash(
+        core.SerializeScript(
+          normalizers.NormalizeScript(inputs[i].cell_output.lock)
+        )
+      )
+      .serializeJson();
+    if (lockSet.has(scriptHash)) {
+      if (groups.get(scriptHash) === undefined) groups.set(scriptHash, []);
+      groups.get(scriptHash)!.push(i);
+    }
+  }
+  return groups;
+}
+
+function calcRawTxHash(tx: TransactionSkeletonType): Reader {
+  return utils.ckbHash(
+    core.SerializeRawTransaction(
+      normalizers.NormalizeRawTransaction(createTransactionFromSkeleton(tx))
+    )
+  );
+}
+
+export interface Hasher {
+  update(message: ArrayBuffer): void;
+  digest(): Hash;
+}
+
+type Group = {
+  index: number;
+  lock: Script;
+  message: Hash;
+};
+
+/**
+ * Return an array of messages as well as their corresponding position indexes and locks for signing a P2PKH transaction.
+ * For more details, please see:
+ * https://github.com/nervosnetwork/ckb-system-scripts/wiki/How-to-sign-transaction
+ *
+ * @param tx TxSkeleton with all input cells' witnessArgs.lock filled with 0.
+ * @param locks Locks you want to sign, e.g. you don't need to sign ACP cells.
+ * @param hasher Message hasher, defaults to CKB blake2b hasher. Check
+ * https://github.com/nervosnetwork/ckb-system-scripts/blob/e975e8b7d5231fdb1c537b830dd934b305492417/c/secp256k1_blake160_sighash_all.c#L22-L28 for more.
+ * @returns An array of Group containing: lock of the input cell you need to sign, message for signing, witness index of this message (first index of the input cell with this lock).
+ */
+export function createP2PKHMessageGroup(
+  tx: TransactionSkeletonType,
+  locks: Script[],
+  hasher?: Hasher
```

Comment:
This function may be extended in the future, and the second parameter should preferably be exposed as an object, otherwise, it will probably look like this when the function needs to be extended

```ts
export function createP2PKHMessageGroup(tx, locks, hasher?, feature1?) {}
export function createP2PKHMessageGroup(tx, locks, hasher?, feature1?, feature2?) {}
export function createP2PKHMessageGroup(tx, locks, hasher?, feature1?, feature2?, ...) {}

```